Description: Enable building when OpenSSL has disabled SSLv2
 Throw exception if SSLv2 connection requested but OPENSSL_NO_SSL2 defined.
 Fall back to returning invalid state if exceptions are disabled.
Author: Lubo-B¹ Dole¾el <lubos@dolezel.info>-A
Bug: 621402, 622027, 622070
Forwarded: no

--- boost1.46-1.46.1.orig/boost/asio/ssl/detail/openssl_context_service.hpp
+++ boost1.46-1.46.1/boost/asio/ssl/detail/openssl_context_service.hpp
@@ -19,6 +19,7 @@
 #include <boost/asio/detail/config.hpp>
 #include <cstring>
 #include <string>
+#include <stdexcept>
 #include <boost/function.hpp>
 #include <boost/asio/error.hpp>
 #include <boost/asio/io_service.hpp>
@@ -66,6 +67,14 @@
   {
     switch (m)
     {
+#if defined(OPENSSL_NO_SSL2)
+    case context_base::sslv2:
+    case context_base::sslv2_client:
+    case context_base::sslv2_server:
+      boost::throw_exception(std::runtime_error("SSLv2 is not supported in your OpenSSL"));
+      impl = ::SSL_CTX_new(0);  // if exceptions are disabled
+      break;
+#else
     case context_base::sslv2:
       impl = ::SSL_CTX_new(::SSLv2_method());
       break;
@@ -75,6 +84,7 @@
     case context_base::sslv2_server:
       impl = ::SSL_CTX_new(::SSLv2_server_method());
       break;
+#endif
     case context_base::sslv3:
       impl = ::SSL_CTX_new(::SSLv3_method());
       break;
